Sources/SwiftFusion/Inference/ValuesStorage.swift

```diff
@@ -17,368 +17,369 @@
 
 import PenguinStructures
 
-// MARK: - Storage for contiguous arrays of `Differentiable` values.
+// MARK: - Algorithms on arrays of `Differentiable` values.
 
-/// Contiguous storage of homogeneous `Differentiable` values of statically unknown type.
-class AnyDifferentiableStorage: AnyArrayStorage {
-  typealias DifferentiableImplementation = AnyDifferentiableStorageImplementation
-  var differentiableImplementation: DifferentiableImplementation {
-    fatalError("implement me!")
-  }
-
-  /// Identifies the values' `TangentVector`.
-  var tangentIdentifier: ObjectIdentifier {
-    differentiableImplementation.tangentIdentifier_
+extension ArrayStorage where Element: Differentiable, Element.TangentVector: EuclideanVectorN {
+  /// Returns the zero `TangentVector`s of the contained elements.
+  var tangentVectorZeros: ArrayBuffer<Element.TangentVector> {
+    withUnsafeMutableBufferPointer { vs in
+      .init(vs.lazy.map { $0.zeroTangentVector })
+    }
   }
 
-  /// The values' zero `TangentVector`s.
-  var zeroTangent: AnyArrayBuffer<AnyVectorStorage> {
-    differentiableImplementation.zeroTangent_
-  }
-  
-  /// Moves `self` along the vector starting at `directionsStart`.
+  /// Moves each element of `self` along the corresponding element of `directions`.
   ///
-  /// Precondition: `directionsStart` points to memory with at least `count` initialized
-  /// `Element.TangentVector`s where `Element` is the element type of `self`.
-  final func move(along directionsStart: UnsafeRawPointer) {
-    differentiableImplementation.move_(along: directionsStart)
+  /// - Requires: `directions.count >= self.count`.
+  func move(along directions: ArrayBuffer<Element.TangentVector>) {
+    withUnsafeMutableBufferPointer { vs in
+      directions.withUnsafeBufferPointer { ds in
+        for i in vs.indices {
+          vs[i].move(along: ds[i])
+        }
+      }
+    }
   }
 }
 
-extension AnyArrayBuffer where Storage: AnyDifferentiableStorage {
-  /// Identifies the values' `TangentVector`.
-  var tangentIdentifier: ObjectIdentifier {
-    storage.tangentIdentifier
+// MARK: - Algorithms on arrays of `EuclideanVectorN` values.
+
+extension ArrayStorage where Element: EuclideanVectorN {
+  /// Adds `others` to `self`.
+  ///
+  /// - Requires: `others.count >= count`.
+  func add(_ others: ArrayBuffer<Element>) {
+    withUnsafeMutableBufferPointer { vs in
```

Comment:
I don't love the code pattern repetition in these three methods; please refactor if you need to do it again.